{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Design of exon capture probes from _Trifolium repens_ transcriptome\n",
    "### **Author:** James S. Santangelo\n",
    "### **Date:** August 13<sup>th</sup>, 2018\n",
    "\n",
    "This pipeline is is meant for designing exon capture probes from transcript sequences using the reference genome of a closely related species. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 1: Downloading transcriptome and reference genome\n",
    "\n",
    "* The white clover transcriptome was downloaded from [BMC Genomics](https://bmcgenomics.biomedcentral.com/articles/10.1186/1471-2164-14-100#Sec19), which also includes the description of the transcriptome generated by Nagy *et al.* (2014).\n",
    "* The *Trifolium pratense* reference genome and annotation files (both *.gtf* and *.gff3* files) were downloaded from [Ensembl plants](https://plants.ensembl.org/info/website/ftp/index.html) (release 40, July 2018) on July 31,<sup>th</sup>, 2018. The genome was published by [De Vega *et al.* (2015)](https://www.nature.com/articles/srep17394) and full assembly details are available on the [European Nucleotide Archive](https://www.ebi.ac.uk/ena/data/view/GCA_900079335.1)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 2: Retrieving exons from the _Trifolium pratense_ reference genome\n",
    "\n",
    "* Prior to any filtering, the _T. pratense_ genome contained **184,849 unique exons** from **39,948 unique genes**\n",
    "* Using the *.gff3* annotation file and the `gffutils` python module, I extracted the sequence of each of these exons and wrote them to a new FASTA file. The first 5 sequences in the exon FASTA are shown below,which represent the first 5 exons from the gene with ID 4341"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">Tp57577_TGAC_v2_mRNA4484.exon1.gene4341\n",
      "CAGCGAATGAACGAGGTTCTCCCACACAGCGAACCGCATTCTCCCTCAGCGAACCCTAAGTCGTTCTTCCTCTCGGTGAAATGCAAATTATGTCCCTCTCATGCTGCGACGAAATCGATTACGGCGACCACGTTCTGCTACCACGGTGAAATTCGTCCTcccccccTTTCAAG\n",
      ">Tp57577_TGAC_v2_mRNA4484.exon2.gene4341\n",
      "AAGTAGTTATGGATCGTAGTTGGATGAATGCTAACCAGTTCAGTGAAGAGTATAAGAAAGAG\n",
      ">Tp57577_TGAC_v2_mRNA4484.exon3.gene4341\n",
      "GCTTGTAATGAATACCACCACATCGTGTCTCGTGGGGGATATGAGTTGATTGAAGATAAGATGTTGCAAGATGAAATTAAACAAAGACAAAAGTCAGCTGGAGATTCCCTACCTACACCTCCATCTTCACCAAAACGCTATGAGAAGTGGATAAGGGGACGAAAAAGCATATGTATCAGAAGGATAAATG\n",
      ">Tp57577_TGAC_v2_mRNA4484.exon4.gene4341\n",
      "aagttgaggctgaaactagaattgacaatgttgagactgaaaTTGATCCTAAAAATGTTGAAACTAGAGGTGACAATGTTGAGACTGAAATTGATTCTGAAAAAGATGGAAATGTTGGTGTTGATGTAAGATATCGGAAGAATTTAGTATACACAAGGAAAAAGAATATCATTCCTGAATnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnTAGTGATTCAATTTCCGAATTTTCTCCtgctccagttgcaaaaatgaatact\n",
      ">Tp57577_TGAC_v2_mRNA4484.exon5.gene4341\n",
      "catcttggaagacttgaggataaagagtgatgaaccaatgaaatttattgtgataataaatctgctattagcat\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "less ../../example_outputs/pratenseExons_unfiltered_First5.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* The unfiltered _T. pratense_ exons then had to be filtered. I imposed the following two filtering criteria:\n",
    "\n",
    "    1) The exons could not contain any ambiguous nucleotides (i.e. 'n' or 'N').\n",
    "    \n",
    "    2) The exons had to be greater than 200 bp in length since evidence suggests that the capture efficiency of probes declines for sequences shorter than 200 bp (see [Bragg *et al.* (2016)](https://www.ncbi.nlm.nih.gov/pubmed/26215687))\n",
    "<br>\n",
    "<br>\n",
    "* Following filtering, there were **72,355 unique exons** remaining from **35,109 unique genes**."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 3: Reciprocal best blast\n",
    "\n",
    "* In this step, I perform a reciprocal best blast between *T. repens* transcripts and the filtered *T. pratense* reference exons.\n",
    "<br>\n",
    "<br>\n",
    "First, I Blasted the transcriptome against the exon sequences using the following command:\n",
    "<br>\n",
    "<br>\n",
    "`blastn -query Transcriptome.fasta -db Exons.db -out <OUTFILE> -outfmt '6 qseqid sseqid pident length mismatch gapopen qstart qend qlen sstart send slen evalue bitscore qcovs qcovhsp' -num_threads 2 -evalue 1e-10 -max_target_seqs 1 -max_hsps 1`\n",
    "<br>\n",
    "<br>\n",
    "`-evalue 1e-10` Keep only hits where Evalue is less than 1e<sup>-10</sup> <br>\n",
    "`-max_target_seq` Ensure that each transcript hits at most 1 exon target (i.e. the best one) <br>\n",
    "`-max_hsps` Ensure that only the top hit for each target sequence is included in the final output. \n",
    "<br>\n",
    "<br>\n",
    "* This approach resulted in **43,714 hits**. The first 10 hits are shown in the dataframe below as an example."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qseqid</th>\n",
       "      <th>sseqid</th>\n",
       "      <th>pident</th>\n",
       "      <th>length</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>gapopen</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>qlen</th>\n",
       "      <th>sstart</th>\n",
       "      <th>send</th>\n",
       "      <th>slen</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>qcovs</th>\n",
       "      <th>qcovhsp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>wcd-8466</td>\n",
       "      <td>Tp57577_TGAC_v2_mRNA3036.exon6.gene2951</td>\n",
       "      <td>97.319</td>\n",
       "      <td>373</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>373</td>\n",
       "      <td>999</td>\n",
       "      <td>372</td>\n",
       "      <td>1</td>\n",
       "      <td>475</td>\n",
       "      <td>1.260000e-180</td>\n",
       "      <td>632</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>wcd-7834</td>\n",
       "      <td>Tp57577_TGAC_v2_mRNA12970.exon3.gene12562</td>\n",
       "      <td>92.000</td>\n",
       "      <td>450</td>\n",
       "      <td>34</td>\n",
       "      <td>2</td>\n",
       "      <td>418</td>\n",
       "      <td>867</td>\n",
       "      <td>999</td>\n",
       "      <td>1</td>\n",
       "      <td>448</td>\n",
       "      <td>472</td>\n",
       "      <td>4.550000e-180</td>\n",
       "      <td>630</td>\n",
       "      <td>45</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>wcd-7404</td>\n",
       "      <td>Tp57577_TGAC_v2_mRNA22327.exon1.gene21589</td>\n",
       "      <td>95.349</td>\n",
       "      <td>129</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>871</td>\n",
       "      <td>999</td>\n",
       "      <td>999</td>\n",
       "      <td>550</td>\n",
       "      <td>422</td>\n",
       "      <td>550</td>\n",
       "      <td>3.270000e-52</td>\n",
       "      <td>206</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>wcd-42747</td>\n",
       "      <td>Tp57577_TGAC_v2_mRNA26352.exon1.gene25485</td>\n",
       "      <td>94.316</td>\n",
       "      <td>563</td>\n",
       "      <td>21</td>\n",
       "      <td>4</td>\n",
       "      <td>403</td>\n",
       "      <td>965</td>\n",
       "      <td>999</td>\n",
       "      <td>639</td>\n",
       "      <td>88</td>\n",
       "      <td>639</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>852</td>\n",
       "      <td>56</td>\n",
       "      <td>56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>wcd-41619</td>\n",
       "      <td>Tp57577_TGAC_v2_mRNA10895.exon4.gene10542</td>\n",
       "      <td>96.035</td>\n",
       "      <td>227</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>479</td>\n",
       "      <td>705</td>\n",
       "      <td>999</td>\n",
       "      <td>227</td>\n",
       "      <td>1</td>\n",
       "      <td>227</td>\n",
       "      <td>1.090000e-101</td>\n",
       "      <td>370</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>wcd-3612</td>\n",
       "      <td>Tp57577_TGAC_v2_mRNA13531.exon1.gene13095</td>\n",
       "      <td>96.070</td>\n",
       "      <td>458</td>\n",
       "      <td>13</td>\n",
       "      <td>3</td>\n",
       "      <td>31</td>\n",
       "      <td>483</td>\n",
       "      <td>999</td>\n",
       "      <td>135</td>\n",
       "      <td>592</td>\n",
       "      <td>592</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>741</td>\n",
       "      <td>45</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>wcd-1589</td>\n",
       "      <td>Tp57577_TGAC_v2_mRNA9035.exon4.gene8741</td>\n",
       "      <td>97.568</td>\n",
       "      <td>329</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>200</td>\n",
       "      <td>528</td>\n",
       "      <td>999</td>\n",
       "      <td>326</td>\n",
       "      <td>1</td>\n",
       "      <td>326</td>\n",
       "      <td>6.050000e-159</td>\n",
       "      <td>560</td>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>wcd-13378</td>\n",
       "      <td>Tp57577_TGAC_v2_mRNA39072.exon1.gene37796</td>\n",
       "      <td>90.845</td>\n",
       "      <td>710</td>\n",
       "      <td>38</td>\n",
       "      <td>12</td>\n",
       "      <td>23</td>\n",
       "      <td>729</td>\n",
       "      <td>999</td>\n",
       "      <td>313</td>\n",
       "      <td>998</td>\n",
       "      <td>998</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>929</td>\n",
       "      <td>71</td>\n",
       "      <td>71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>wcc-26375</td>\n",
       "      <td>Tp57577_TGAC_v2_mRNA27039.exon1.gene26146</td>\n",
       "      <td>95.929</td>\n",
       "      <td>1007</td>\n",
       "      <td>30</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>999</td>\n",
       "      <td>999</td>\n",
       "      <td>1505</td>\n",
       "      <td>502</td>\n",
       "      <td>1625</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1624</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>wcc-14061</td>\n",
       "      <td>Tp57577_TGAC_v2_mRNA17836.exon6.gene17237</td>\n",
       "      <td>96.082</td>\n",
       "      <td>485</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>180</td>\n",
       "      <td>664</td>\n",
       "      <td>999</td>\n",
       "      <td>1</td>\n",
       "      <td>485</td>\n",
       "      <td>485</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>800</td>\n",
       "      <td>49</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      qseqid                                     sseqid  pident  length  \\\n",
       "0   wcd-8466    Tp57577_TGAC_v2_mRNA3036.exon6.gene2951  97.319     373   \n",
       "1   wcd-7834  Tp57577_TGAC_v2_mRNA12970.exon3.gene12562  92.000     450   \n",
       "2   wcd-7404  Tp57577_TGAC_v2_mRNA22327.exon1.gene21589  95.349     129   \n",
       "3  wcd-42747  Tp57577_TGAC_v2_mRNA26352.exon1.gene25485  94.316     563   \n",
       "4  wcd-41619  Tp57577_TGAC_v2_mRNA10895.exon4.gene10542  96.035     227   \n",
       "5   wcd-3612  Tp57577_TGAC_v2_mRNA13531.exon1.gene13095  96.070     458   \n",
       "6   wcd-1589    Tp57577_TGAC_v2_mRNA9035.exon4.gene8741  97.568     329   \n",
       "7  wcd-13378  Tp57577_TGAC_v2_mRNA39072.exon1.gene37796  90.845     710   \n",
       "8  wcc-26375  Tp57577_TGAC_v2_mRNA27039.exon1.gene26146  95.929    1007   \n",
       "9  wcc-14061  Tp57577_TGAC_v2_mRNA17836.exon6.gene17237  96.082     485   \n",
       "\n",
       "   mismatch  gapopen  qstart  qend  qlen  sstart  send  slen         evalue  \\\n",
       "0         9        1       1   373   999     372     1   475  1.260000e-180   \n",
       "1        34        2     418   867   999       1   448   472  4.550000e-180   \n",
       "2         6        0     871   999   999     550   422   550   3.270000e-52   \n",
       "3        21        4     403   965   999     639    88   639   0.000000e+00   \n",
       "4         9        0     479   705   999     227     1   227  1.090000e-101   \n",
       "5        13        3      31   483   999     135   592   592   0.000000e+00   \n",
       "6         5        2     200   528   999     326     1   326  6.050000e-159   \n",
       "7        38       12      23   729   999     313   998   998   0.000000e+00   \n",
       "8        30        9       1   999   999    1505   502  1625   0.000000e+00   \n",
       "9        19        0     180   664   999       1   485   485   0.000000e+00   \n",
       "\n",
       "   bitscore  qcovs  qcovhsp  \n",
       "0       632     37       37  \n",
       "1       630     45       45  \n",
       "2       206     13       13  \n",
       "3       852     56       56  \n",
       "4       370     23       23  \n",
       "5       741     45       45  \n",
       "6       560     33       33  \n",
       "7       929     71       71  \n",
       "8      1624    100      100  \n",
       "9       800     49       49  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "colnames = [\"qseqid\", \"sseqid\", \"pident\", \"length\",\n",
    "            \"mismatch\", \"gapopen\", \"qstart\", \"qend\",\n",
    "            \"qlen\", \"sstart\", \"send\", \"slen\", \"evalue\",\n",
    "            \"bitscore\", \"qcovs\", \"qcovhsp\"]\n",
    "\n",
    "transcriptome_to_exons = pd.read_table(\"../../image/recipBlast_Trans2Exons_First10.txt\",\n",
    "                                       sep=\"\\t\",\n",
    "                                       names=colnames)\n",
    "transcriptome_to_exons"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* I then used the same approach to BLAST the filtered reference exons against the transcript sequences, which resulted in **39,973 hits**.\n",
    "* I identified reciprocal hits by matching the query and subject sequences from the 2 independent BLAST results. A hit was identified as reciprocal if the query sequence/subject sequence pairs were the same in both datasets (i.e. a given exon is the best match to a given transcript in the first BLAST and that same transcript is the best match to the same exon in the second BLAST).Using this approach, I identified **22,121 reciprocal hits**.\n",
    "* I then filtered these reciprocal hits using the following criteria:\n",
    "\n",
    "    1) For the same reason as above, the length of the alignment had to be > 200 bp.\n",
    "    \n",
    "    2) No two exons could come from the same gene\n",
    "        * Where a gene had more than one exon, I selected the one with the highest bitscore. If more than one exon had the same bitscore, I selected the one with the greatest % identity to the transcript sequence.  \n",
    "    \n",
    "    3) The % identity of the transcript-exon alignment had to be greater than 80%.\n",
    "    \n",
    "    4) More than 50% of the exon had to be covered by transcript sequence. \n",
    "    \n",
    "* Following filtering, **10,176 reciprocal hits** remained. These are the hits that were used to extract sequences from the transcriptome for which capture probes will be designed."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 3: Sequences for probe design\n",
    "\n",
    "* The sequences for probe design were extracted from the _T. repens_ transcriptome using the query sequence start and end coordinates from the reciprocal hits data frame, which is identical to the table shown above but has been filtered. \n",
    "* Importantly, I took care to ensure that strandedness was maintained.\n",
    "    * For sequences on the + strand (i.e. start coordinate < end coordinate), the sequences could be extracted directly from the transcriptome using these coordinates. \n",
    "    * For sequences on the â€“ strand (i.e. end coordinate < start coordinate), the start and end coordinates had to be swapped and the resulting sequence was reverse complemented. \n",
    "* While extracting the sequence, I additionally extracted the transcript annotation and place this in the FASTA header for easy querying (see below).\n",
    "* Finally, I only kept sequences if their % GC content was between 30% and 70% since sequences with low or high GC content have reduced capture efficiency (see [Bi _et al._ (2016)](https://bmcgenomics.biomedcentral.com/articles/10.1186/1471-2164-13-403)). The final, filtered FASTA file has **10,121 sequences**, covering approximately **5.29 Mbp** of sequence. The first 5 sequences from the FASTA file are shown below. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">Transcript_id:wcd-8855;Strand:-;PratenseExon_id:Tp57577_TGAC_v2_mRNA4492.exon1.gene4349;Length:1033;Transcript_annot:no annotation\n",
      "GAAGCTAAATACATGATACTAAAAAATCTGATGGAAGAAAAACAGTTGGATTTTAATCAGCCACTTTTATCAGTAAGGCGATTCCCATCAACAGCACCATACTGAAAGACGACAAGTAAAAGGAAAACCGATAAGCCAATATCGAAGAGAACTCGTCTTCCTGCTTACAAATCAGAAGTTAAAWWCAGGTCCGGTGACTAACCCTGGAACTGTTCTTTCGTATGGGAGAAGACTCCCGGAAGACCTATAGATGAAAGCAAAGYKTCAACAAYGGCTATYGAAGGACCTCTYGTCGCTCCTAAGCTTCCACCKGGAAGGGTTTTAAAAGTTGAACAGCAAGATTTTGATAAAACTCCTAAGGGCTCATTGGTTACTCGGTCCAGAACAGAAAGCACTGTTTCCAACTCTATGAGTATTGCATCTTTGGATAGTAAAGAAGCAAATCATGACAGCCGAAAAGAAGAGATTCTGGAAAAGGAAAGTTCTGGTTCAGATAATGGGGATGAGACCTATCTAGATGCTCGTGATACACTTTCTAGAACCGAATCATTCTTCATGAACTGCAGTGTGAGTGGTATGAGTGGATGTGATGATCGAGAGGTTCAACCATCCGAAACTTTTTCGGCAGATCAACAGGCTCGTGATTTCATGATCGGTAGGTTTTTGCCTGCAGCAAAGGCTATGGCTTCTGAAACGCCTCATGTCCAATACGCCTCTAGGAAGCCAATTGTTAGACAGGAACAACCTAGACTAGTAAGGAAAGTAGAAAGTGGAGCCAAGTCCCGTCCCTTGGACCCGAAATGGCAAAAGGTTTTGCCACATTATGCACAAGATACGAGCCTCAATGAAAGTGAAGACGAAAGCGACGAAGATAGATACGAAAGCCATGCACCCAAAGTTTGTGGTTTATTTCCTCGGTTCTGCCTTTTAAGTCCATTGCAAGGACTAAGAGTGGAGGATAAAATTGTAAATTCTGCAATTCATGGAGTGCAGGGTAAATCGATTGTTTCTCRTAGAAAAAAATGCRAAAGAG\n",
      ">Transcript_id:wcc-36701;Strand:+;PratenseExon_id:Tp57577_TGAC_v2_mRNA4384.exon3.gene4248;Length:203;Transcript_annot:protein kinase family protein\n",
      "GAGAAAAGACTCAAGAAGCATTTTGATAGCATTCCGACTTTGGATGAGCTTCAAGCTTTAGGTGGAGAGGGTTTTAGGGCTGATGTCATTGTTGTGGATGCAAAGAAAGATAGAAAGTTATCTATGTTGAAGCAACTAATTGTCACATTGGTTAAAGGGTTGAACGCAATCTAGCTGCAATAATTAAGAAGATAGCTGTATTA\n",
      ">Transcript_id:wcd-1293;Strand:-;PratenseExon_id:Tp57577_TGAC_v2_mRNA4429.exon3.gene4289;Length:707;Transcript_annot:af307095_1non-cell-autonomous protein pathway2\n",
      "GATTTCCTGGTGACCTCAAAGTAACAGTGAGTTACATTTTAGAAAAAAACTCTCTAACCATACTTATGCAAGCAAAAGCTTTAAATAAACCTACACCTGTTAATCTAGTTAACCATGCTTATTGGAACCTAGGAAACCACAATAGTGGCAATATTTTAGATGAAGTGGTTCAAATCTTTGGATCCAAAAACACAGTCTTTGATGACAACCTCATTCCAACTGGAAAAATTTCATCAGTCAAAGGAACACCAAATGATTTTCTAAAACCACAAATTGTTGGTACAAGAATTAATCAATTGCCAAAAACCAATGGTTATAATGTTAATTATGTGCTTAATAAAGGAAAATTAGGTAACAAAGAAGAATTAAAGGTTGCGGCTATTGTGTTCGATAAGAAATCAGGGAGAGTTATGAAGCTTAGTACTAATGCTCCTGGATTGCAATTTTATACCGCTAATTTTGTAAAGAATGATAAAGGAAAAGGTGGATTTGTTTATCAACCACGTAGCGCGCTTTGTTTGGAGAGTCAAGCTTTTCCTGATTCTGTTAATCATCCTAATTTTCCTTCTACTATTGTTACAAAAGAAAAGCCTTATAAGCATGTCATGTTGTTGAAGTTTTCAACCAAAGTTCCAGATGCTTTTTCACAATTCTAGTGCTTAGCATGATTCTTGTGTTCATACAAACTATGAAGTAAAATAAGTGTT\n",
      ">Transcript_id:wcc-18245;Strand:+;PratenseExon_id:Tp57577_TGAC_v2_mRNA4423.exon5.gene4283;Length:213;Transcript_annot:no annotation\n",
      "GTCCTGAGTTGAGGTTAGCAAAAGTTAAGGTTGATGCTGATTTCACTGCTCCTTGCAACCCTTCTATCTTATATCCTGAGGATGGTGGAAACTTGCACTGTTTCACAGCAGTGACAGCTTGTGCAGTTCTTGATGTTCTTGGTCCTCCATATTCTGATTTTGATGGCAGACACTGCACATACTATACTAATTACCCTTTTCCCAACTTTTCAG\n",
      ">Transcript_id:wcc-26763;Strand:+;PratenseExon_id:Tp57577_TGAC_v2_mRNA4458.exon5.gene4317;Length:213;Transcript_annot:f-box family protein\n",
      "TTCTTGTTTGCGGTGCTTGAGAGGTCACAATGGTCCAGTGTTATCACTATCAAATAAATTATTAGGAGAGGATGGCAGCAAAGTCTTAGCAAGTGGAGGAGAAGATGGTACTGTTCGACTATGGTCCCTTAGCTCAAGTGGAAAAAGAGGGCAGCGTGCTTTGAAGGCTACACTCTATGGACATAGAAAACCTATAAATTTGATGTCAGTTTC\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "less ../../image/exonTargets_First5.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Transcript annotations\n",
    "\n",
    "* Because we are interested in including sequences underlying proteins that may be influenced by urbanization, I queried the transcript sequence annotations for certain abiotic/biotic factor that may be influenced by cities. Specifically, I looked for:\n",
    "    * salt (6 hits)\n",
    "    * drought (2 hits)\n",
    "    * nodul* (35 hits)\n",
    "    * cyanog* (2 hits): A cyanogenic beta-glucosidase and a non-cyanogenic beta-glucosidase\n",
    "    * cold (1 hit)\n",
    "    * stress (22 hits)\n",
    "    * dehydrin (involved in drought response, 1 hit)\n",
    "    \n",
    "The above queries resulted in a combined **68 hits**. \n",
    "\n",
    "Also, note that **1,642 sequences** have no annotation. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Additional considerations\n",
    "\n",
    "1) We still have ~0.7 Mbp worth of sequence for which we can design probes since the kit we are buying comes in a 6 Mbp size. How should we target additional sequences?\n",
    "\n",
    "* First, we can add sequences for regions we are interested in using the white clover EST/transcript database. For example, only linamarase was a hit to the transcriptome but I suggest we design probes for all _CYP79D15_ and _Li_ sequences that are currently published, including flanking regions. However, this will only be a fraction of the 0.7 Mbp. \n",
    "* We could also just design 200, 400, 600 and 800 bp probes for random stretches of transcript sequences, targeting only those that were not homologous to red clover exons. While not homologous to red clover exons, these transcripts still represent sequences being expressed by white clover and may be sequences unique to white clover. While some of these probes risk overlapping exon boundaries and not binding, I don't think this will be much of an issue.\n",
    "\n",
    "2) Should we BLAST the final set of exons against itself to check for repetitive segments? My sense is that we don't have to since the probe design company screens for repeats as part of their pipeline. "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
